### PR TITLE
Push metrics bridge

### PR DESCRIPTION
**Pull request by johnjmartin**
## Description 

This adds the ability for a bridge node operator to push metrics to an external metrics proxy for aggregation. The implementation is mostly copied from https://github.com/MystenLabs/sui/blob/main/crates/sui-node/src/metrics.rs

## Test plan 

Tested in testnet

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

**Review by longbowlu — COMMENTED** [review]
would you mind rebasing this to https://github.com/MystenLabs/sui/pull/18877 for better readability?

File: crates/sui-bridge/src/config.rs [hunk 1]
```diff
@@ -117,12 +117,23 @@ pub struct BridgeNodeConfig {
     /// Network key used for metrics pushing
     #[serde(default = "default_ed25519_key_pair")]
     pub metrics_key_pair: NetworkKeyPair,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub metrics: Option<MetricsConfig>,
```
Comment (longbowlu) [inline crates/sui-bridge/src/config.rs:121]: let's call this `metrics_config`

File: crates/sui-bridge/src/main.rs [hunk 2]
```diff
@@ -48,6 +49,11 @@ async fn main() -> anyhow::Result<()> {
     let metadata =
         BridgeNodePublicMetadata::new(VERSION.into(), config.metrics_key_pair.public().clone());
 
+    start_metrics_push_task(
+        &config.metrics,
+        config.metrics_key_pair.copy(),
```
Comment (longbowlu) [inline crates/sui-bridge/src/main.rs:54]: do we need `copy` here?

File: crates/sui-bridge/src/metrics.rs [hunk 3]
```diff
@@ -15,6 +20,140 @@ const FINE_GRAINED_LATENCY_SEC_BUCKETS: &[f64] = &[
...
+    certificate: std::sync::Arc<sui_tls::SelfSignedCertificate>,
+    client: reqwest::Client,
+}
+
+impl MetricsPushClient {
+    pub fn new(metrics_key: sui_types::crypto::NetworkKeyPair) -> Self {
+        use fastcrypto::traits::KeyPair;
+        let certificate = std::sync::Arc::new(sui_tls::SelfSignedCertificate::new(
+            metrics_key.private(),
+            sui_tls::SUI_VALIDATOR_SERVER_NAME,
```
Comment (longbowlu) [inline crates/sui-bridge/src/metrics.rs:33]: should we use different values?

File: crates/sui-bridge/src/metrics.rs [hunk 4]
```diff
@@ -15,6 +20,140 @@ const FINE_GRAINED_LATENCY_SEC_BUCKETS: &[f64] = &[
...
+/// Starts a task to periodically push metrics to a configured endpoint if a metrics push endpoint
+/// is configured.
+pub fn start_metrics_push_task(
+    metrics_config: &Option<MetricsConfig>,
+    metrics_key_pair: NetworkKeyPair,
+    registry: RegistryService,
+) {
+    use fastcrypto::traits::KeyPair;
+
+    const DEFAULT_METRICS_PUSH_INTERVAL: Duration = Duration::from_secs(60);
```
Comment (longbowlu) [inline crates/sui-bridge/src/metrics.rs:65]: is this the default value we use for validator node? it seems a bit large

File: crates/sui-bridge/src/config.rs [hunk 5]
```diff
@@ -117,12 +117,23 @@ pub struct BridgeNodeConfig {
...
     get_key_pair_from_rng(&mut rand::rngs::OsRng).1
 }
 
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct MetricsConfig {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub push_interval_seconds: Option<u64>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub push_url: Option<String>,
```
Comment (longbowlu) [inline crates/sui-bridge/src/config.rs:134]: i think `push_url` should be required? if the node does not want to push they should set `MetricsConfig` as `None`

File: crates/sui-bridge/src/metrics.rs [hunk 6]
```diff
@@ -15,6 +20,140 @@ const FINE_GRAINED_LATENCY_SEC_BUCKETS: &[f64] = &[
...
+                .unwrap_or(DEFAULT_METRICS_PUSH_INTERVAL);
+            let url = reqwest::Url::parse(url).expect("unable to parse metrics push url");
+            (interval, url)
+        }
+        _ => return,
+    };
+
+    let mut client = MetricsPushClient::new(metrics_key_pair.copy());
+
+    async fn push_metrics(
```
Comment (longbowlu) [inline crates/sui-bridge/src/metrics.rs:83]: this seems to be a duplicate of https://github.com/MystenLabs/sui/blob/064cb642e0ac3f1d12abb750041cebad8edfee82/crates/sui-node/src/metrics.rs#L75. I think we can pull it out and put it somewhere for reuse. 
`mysten-common` feel like a candidate crate

File: crates/sui-bridge/src/metrics.rs (hunk 3 above)
Comment (johnjmartin) [reply crates/sui-bridge/src/metrics.rs:33]: Not unless we run a dedicated metrics proxy for the bridge: https://github.com/MystenLabs/sui/blob/d275762040247747ef429eda6309ffbef005c28b/crates/sui-proxy/src/admin.rs#L260-L263

File: crates/sui-bridge/src/metrics.rs (hunk 4 above)
Comment (johnjmartin) [reply crates/sui-bridge/src/metrics.rs:65]: It is, from mysten validators configs:
```
metrics:
  push-interval-seconds: 60
  ```

File: crates/sui-bridge/src/config.rs (hunk 1 above)
Comment (johnjmartin) [reply crates/sui-bridge/src/config.rs:121]: `metrics` is what it is called in `sui-node.yaml` validator configs. Probably better to keep them the same?

File: crates/sui-bridge/src/main.rs (hunk 2 above)
Comment (johnjmartin) [reply crates/sui-bridge/src/main.rs:54]: I believe so, `config` is also passed to `run_bridge_node(config,` on the lines below this, so without the copy I get a `value used here after partial move` compile error

File: crates/sui-bridge/src/metrics.rs (hunk 6 above)
Comment (johnjmartin) [reply crates/sui-bridge/src/metrics.rs:83]: That is a great idea, seems easy to do. Do you mind if I do that in a follow-up PR?

Comment (longbowlu) [reply crates/sui-bridge/src/metrics.rs:83]: of course! let's add a TODO here

**Review by longbowlu — APPROVED** [review]

**Pull request merged**